### PR TITLE
Function pointers in IELE

### PR DESCRIPTION
**Pull request by dwightguth**
Per our previous discussion, this PR adds function pointers to IELE by allowing global function names anywhere an operand can be used in the assembler, and adding semantics for dynamic call targets computed from an integer function offset.

### CODE REVIEW

**Comment by nishantjr** [comment]
@dwight possible to add a test for storing a function pointer and retrieving it from memory?
Something along the lines of

```
define @test() {
   %func = @f1
   store %func, 1234
   %loadedFunc = load 1234
   call %loadedFunc()
}

define @f1() {
  ret 42
}
```

**Comment by dwightguth** [comment]
@nishantjr added

File: compiler/src/IeleAssembler.hs [hunk 1]
```diff
@@ -112,8 +112,11 @@ asm_iele_opcode_li opLi = case opLi of
 asm_iele_opcode_call :: IeleOpcodeCall Word16 Word16 -> Put
 asm_iele_opcode_call opCall = case opCall of
   CALL call nargs nreturn -> putWord8 0xf2 >> putWord16be call >> putArgs16 nargs >> putRets16 nreturn
+  CALLDYN nargs nreturn -> putWord8 0xf3 >> putArgs16 nargs >> putRets16 nreturn
   STATICCALL call nargs nreturn -> putWord8 0xf5 >> putWord16be call >> putArgs16 nargs >> putRets16 nreturn
+  STATICCALLDYN nargs nreturn -> putWord8 0xf9 >> putArgs16 nargs >> putRets16 nreturn
   LOCALCALL call nargs nreturn -> putWord8 0xf8 >> putWord16be call >> putArgs16 nargs >> putRets16 nreturn
+  LOCALCALLDYN nargs nreturn -> putWord8 0xf4 >> putArgs16 nargs >> putRets16 nreturn
```
Comment (bmmoore) [inline compiler/src/IeleAssembler.hs]: Why not have f4 for staticcalldyn, nearer to f5 for staticcall and f9 for localcalldyn, near to f8 for localcall

File: compiler/src/IeleDesugar.hs [hunk 2]
```diff
@@ -44,8 +44,8 @@ expandImmediates inst =
   in reverse loads++[inst']
 
 desugarFundef :: Map GlobalName Integer
-              -> FunctionDefinitionP
```
Comment (bmmoore) [inline compiler/src/IeleDesugar.hs]: Could the type synonym be changed so it is still usable here?

**Pull request merged**